File: packages/typescript/src/parallel-executor.ts
/**
 * Parallel Executor — runs multiple pipeline stages concurrently.
 */

import type { Agent } from "./agent.js";
import type {
  AgentRole,
  PipelineStage,
  SharedContext,
  Task,
  TaskResult,
} from "./types.js";

export interface ParallelGroup {
  stages: PipelineStage[];
}

/**
 * Identify groups of consecutive stages that can run in parallel.
 * Stages with `parallel: true` that are adjacent form a group.
 */
export function identifyParallelGroups(
  stages: PipelineStage[]
): (PipelineStage | ParallelGroup)[] {
  const result: (PipelineStage | ParallelGroup)[] = [];
  let currentGroup: PipelineStage[] = [];

  for (const stage of stages) {
    if ((stage as PipelineStage & { parallel?: boolean }).parallel) {
      currentGroup.push(stage);
    } else {
      if (currentGroup.length > 0) {
        result.push({ stages: currentGroup });
        currentGroup = [];
      }
      result.push(stage);
    }
  }

  if (currentGroup.length > 0) {
    result.push({ stages: currentGroup });
  }

  return result;
}

/**
 * Check if an item is a parallel group.
 */
export function isParallelGroup(
  item: PipelineStage | ParallelGroup
): item is ParallelGroup {
  return "stages" in item && Array.isArray((item as ParallelGroup).stages);
}

/**
 * Execute a group of stages in parallel using Promise.allSettled.
 */
export async function executeParallelGroup(
  group: ParallelGroup,
  agents: Map<AgentRole, Agent>,
  task: Task,
  context: SharedContext
): Promise<TaskResult[]> {
  const promises = group.stages.map(async (stage) => {
    const agent = agents.get(stage.role);
    if (!agent) {
      if (stage.required) {
        throw new Error(`Required agent role "${stage.role}" is not registered.`);
      }
      return null;
    }
    return agent.execute(task, context);
  });

  const settled = await Promise.allSettled(promises);
  const results: TaskResult[] = [];
  const errors: Error[] = [];

  for (let i = 0; i < settled.length; i++) {
    const outcome = settled[i];
    if (outcome.status === "fulfilled" && outcome.value !== null) {
      results.push(outcome.value);
    } else if (outcome.status === "rejected") {
      const stage = group.stages[i];
      if (stage.required) {
        errors.push(
          outcome.reason instanceof Error
            ? outcome.reason
            : new Error(String(outcome.reason))
        );
      }
    }
  }

  if (errors.length > 0) {
    throw new AggregateError(
      errors,
      `${errors.length} required parallel stage(s) failed`
    );
  }

  return results;
}

File: packages/python/src/tcs_bpi_ai_team/parallel_executor.py
"""Parallel Executor — runs multiple pipeline stages concurrently."""

from __future__ import annotations

import asyncio
from dataclasses import dataclass, field
from typing import TYPE_CHECKING

from tcs_bpi_ai_team.types import (
    AgentRole,
    PipelineStage,
    SharedContext,
    Task,
    TaskResult,
)

if TYPE_CHECKING:
    from tcs_bpi_ai_team.agent import Agent


@dataclass
class ParallelGroup:
    """A group of pipeline stages that can run concurrently."""

    stages: list[PipelineStage] = field(default_factory=list)


def identify_parallel_groups(
    stages: list[PipelineStage],
) -> list[PipelineStage | ParallelGroup]:
    """Identify groups of consecutive stages that can run in parallel.

    Stages with a `parallel` attribute set to True that are adjacent form a group.
    """
    result: list[PipelineStage | ParallelGroup] = []
    current_group: list[PipelineStage] = []

    for stage in stages:
        if getattr(stage, "parallel", False):
            current_group.append(stage)
        else:
            if current_group:
                result.append(ParallelGroup(stages=current_group))
                current_group = []
            result.append(stage)

    if current_group:
        result.append(ParallelGroup(stages=current_group))

    return result


def is_parallel_group(item: PipelineStage | ParallelGroup) -> bool:
    """Check if an item is a parallel group."""
    return isinstance(item, ParallelGroup)


async def execute_parallel_group(
    group: ParallelGroup,
    agents: dict[AgentRole, Agent],
    task: Task,
    context: SharedContext,
) -> list[TaskResult]:
    """Execute a group of stages in parallel using asyncio.gather."""

    async def run_stage(stage: PipelineStage) -> TaskResult | None:
        agent = agents.get(stage.role)
        if agent is None:
            if stage.required:
                msg = f'Required agent role "{stage.role}" is not registered.'
                raise ValueError(msg)
            return None
        return await agent.execute(task, context)

    tasks = [run_stage(stage) for stage in group.stages]
    results_raw = await asyncio.gather(*tasks, return_exceptions=True)

    results: list[TaskResult] = []
    errors: list[Exception] = []

    for i, outcome in enumerate(results_raw):
        if isinstance(outcome, BaseException):
            if group.stages[i].required:
                err = outcome if isinstance(outcome, Exception) else Exception(str(outcome))
                errors.append(err)
        elif isinstance(outcome, TaskResult):
            results.append(outcome)

    if errors:
        msg = f"{len(errors)} required parallel stage(s) failed"
        raise ExceptionGroup(msg, errors)

    return results
